formats/json-okio/commonMain/src/kotlinx/serialization/json/okio/OkioStreams.kt

```diff
@@ -45,54 +44,47 @@ public fun <T> Json.encodeToSink(
  * @throws [okio.IOException] If an I/O error occurs and sink can't be written to.
  */
 @ExperimentalSerializationApi
-public inline fun <reified T> Json.encodeToSink(
+public inline fun <reified T> Json.encodeToBufferedSink(
     value: T,
-    target: Sink
-): Unit = encodeToSink(serializersModule.serializer(), value, target)
+    target: BufferedSink
+): Unit = encodeToBufferedSink(serializersModule.serializer(), value, target)
 
 
 /**
  * Deserializes JSON from [source] using UTF-8 encoding to a value of type [T] using [deserializer].
  *
- * If [source] is not a [BufferedSource] then called [Source.buffer] for it to create buffered wrapper.
- *
  * Note that this functions expects that exactly one object would be present in the source
  * and throws an exception if there are any dangling bytes after an object.
  *
  * @throws [SerializationException] if the given JSON input cannot be deserialized to the value of type [T].
  * @throws [okio.IOException] If an I/O error occurs and source can't be read from.
  */
 @ExperimentalSerializationApi
-public fun <T> Json.decodeFromSource(
+public fun <T> Json.decodeFromBufferedSource(
     deserializer: DeserializationStrategy<T>,
-    source: Source
+    source: BufferedSource
 ): T {
-    val buffered = if (source is BufferedSource) source else source.buffer()
-    return decodeByReader(deserializer, OkioSerialReader(buffered))
+    return decodeByReader(deserializer, OkioSerialReader(source))
 }
 
 /**
  * Deserializes the contents of given [source] to the value of type [T] using UTF-8 encoding and
  * deserializer retrieved from the reified type parameter.
  *
- * If [source] is not a [BufferedSource] then called [Source.buffer] for it to create buffered wrapper.
- *
  * Note that this functions expects that exactly one object would be present in the stream
```

Comment:
This limitation arised from original `decodeFromStream` function due to the overbuffering. Since we already accepting here `BufferedSource`, and anyone can freely continue to use this source after us, maybe we should reconsider this limitation for this particular function

Reply:
Now it's hard to rewrite it, because the character buffer in `ReaderJsonLexer` is used. As part of another task, it will be possible to work out this optimization. 

Also, the `ReaderJsonLexer` itself can be optimized if moved `CharArrayPool` to the multiplatform code.